### PR TITLE
Add instances for CPS transformers.

### PR DESCRIPTION
**Pull request by sellout**
Conditionalized on transformers >= 0.5.6.

Unfortunately, the operations on CPS transformers are a bit
overconstrained (e.g., Monoid instead of Semigroup), so the instances are
correspondingly more constrained than the Lazy and Strict ones.

### CODE REVIEW

**Comment by RyanGlScott** [comment]
I see your point about the operations on CPS transformers being "overconstrained". Having to impose `Monoid` constraints definitely stings, especially since being able to relax that constraint to `Semigroup` in other places is one of the big appeals behind `semigroupoids` _vis-à-vis_ transformers.

I wonder: is there some function (or set of functions) that `transformers` could define that would allow us to define these instances with `Semigroup` instances?

**Comment by sellout** [comment]
I _think_ the constraint on, say, `CPS.writerT` and `CPS.rwsT` (the biggest culprits) could be weakened directly. But I haven't tried it.

**Comment by sellout** [comment]
I did some poking at `transformers`, reduced the constraints on a bunch of stuff, but nothing that let me reduce the constraint on the operations in semigroupoids so far.

The `Monoid` constraint is pretty fundamental to the CPS version of `WriterT`, though, since you need `mempty` for the initial value of the `w -> m (a, w)` function. Trying to see how much it can be deferred, though.

For a bit more context, I was trying to use `<!>` on an `ExceptT`, where the nested `m` is `CPS.WriterT w IO`. I had originally written a non-polymorphic version of `<!>` since the lack of `Apply (CPS.WriterT w m)` prevented me from using `Apply`'s. I.e., in my use case, I don't care about the `Monoid` constraint, since I'm already using `CPS.WriterT`, which requires it anyway. If I needed to relax that constraint, I could switch to `Strict.` or `Lazy.WriterT`. I guess I could switch anyway, but we use the CPS version pretty consistently.

tl;dr: I don't think any changes to transformers should hold up this change.

**Comment by RyanGlScott** [comment]
> tl;dr: I don't think any changes to transformers should hold up this change.

To be clear, I wouldn't expect you to patch `transformers` itself (at least, not in the short term). If we _could_ find some operation to add that would permit a `Semigroup` constraint, then one possible hack we could add in the interim are functions that expose the internals of CPS transformers, e.g.,

```hs
unWriterT :: WriterT w m a -> w -> m (a, w)
unWriterT = unsafeCoerce
``` 

And then build the operations in `semigroupoids` on top of those functions. However, it sounds like that might not even work. If so, that's rather alarming, since the [Haddocks for the CPS-based `WriterT`](https://hackage.haskell.org/package/transformers-0.5.6.2/docs/Control-Monad-Trans-Writer-CPS.html#t:WriterT) explicitly state that it should be a drop-in replacement for the non-CPS'd `WriterT`! We might have discovered some fundamental limitation which disproves that statement.

**Comment by sellout** [comment]
Yeah, I think it's correct that it's not a drop-in replacement for `Strict.WriterT`. Stuck with `Monoid w` as far as I can discern (and considering that the rest of the transformers are properly constrained on `Semigroup w`, I'm guessing it's not just my own lack of vision).

**Comment by RyanGlScott** [comment]
I had some time to look at this today, and I realized that we can simply adopt the implementations of the methods in the existing `Applicative`, `Alternative`, and `Monad` instances for CPS'd `WriterT` and `RWST` without too much hassle. What's more, none of these instances require a `Monoid` (or even a `Semigroup`) constraint. With this in mind, what about implementing these instances like so?

```hs
instance Alt f => Alt (CPS.WriterT w f) where
  m <!> n = mkWriterT $ \w -> unWriterT m w <!> unWriterT n w

instance Alt f => Alt (CPS.RWST r w s f) where
  m <!> n = mkRWST $ \ r s w -> unRWST m r s w <!> unRWST n r s w

instance Bind m => Apply (CPS.WriterT w m) where
  mf <.> mx = mkWriterT $ \w ->
    unWriterT mf w >>- \(f, w') -> unWriterT (f <$> mx) w'
  -- Or, equivalently:
  -- (<.>) = apDefault

instance Bind m => Apply (CPS.RWST r w s m) where
  mf <.> mx = mkRWST $ \ r s w ->
    unRWST mf r s w >>- \(f, s', w') -> unRWST (f <$> mx) r s' w'
  -- Or, equivalently:
  -- (<.>) = apDefault

instance Bind m => Bind (CPS.WriterT w m) where
  m >>- k = mkWriterT $ \ w ->
    unWriterT m w >>- \(a, w') -> unWriterT (k a) w'

instance Bind m => Bind (CPS.RWST r w s m) where
  m >>- k = mkRWST $ \ r s w ->
    unRWST m r s w >>- \(a, s', w') -> unRWST (k a) r s' w'

instance Plus f => Plus (CPS.WriterT w f) where
  zero = mkWriterT $ const zero

instance Plus f => Plus (CPS.RWST r w s f) where
  zero = mkRWST $ \ _ _ _ -> zero
```

Where:

```hs
mkWriterT :: (w -> m (a, w)) -> CPS.WriterT w m a
mkWriterT = unsafeCoerce

unWriterT :: CPS.WriterT w m a -> w -> m (a, w)
unWriterT = unsafeCoerce

mkRWST :: (r -> s -> w -> m (a, s, w)) -> CPS.RWST r w s m a
mkRWST = unsafeCoerce

unRWST :: CPS.RWST r w s m a -> r -> s -> w -> m (a, s, w)
unRWST = unsafeCoerce
```

I discovered that the absence of the four operations above has already been reported on the `transformers` bug tracker [here](https://hub.darcs.net/ross/transformers/issue/67).

**Comment by sellout** [comment]
Ok, wow, yeah. I was too stuck on the "drop-in replacement for `Strict.WriterT`" (which it's still not, right?) that I figured the implementation should be the same as the `Strict` instances for semigroupoids. and was only trying to "fix" those methods.

This also makes me feel like my changes to transformers are more useful, since it reduces the constraints in some cases (that I figured weren't really helpful when it'd need a `Monoid` eventually anyway).

So ... should I update my PR or do you want to submit that one?

**Comment by RyanGlScott** [comment]
By all means, please update the PR if you're inclined to do so. I could submit my own patch, but I can't make any promises about when I'd get around to doing so.

> I was too stuck on the "drop-in replacement for Strict.WriterT" (which it's still not, right?)

Correct, it's still not a drop-in replacement. In particular, notice that the `Apply` instances that I propose in https://github.com/ekmett/semigroupoids/pull/94#issuecomment-611510787 have `Bind` superclasses. On the other hand, the `Apply` instances for the non-CPS'd `WriterT` and `RWST` only have `Apply` constraints. This is fine, since it parallels a similar discrepancy in the `Applicative` instances in `transformers` (the CPS'd transformers' `Applicative` instances have `Monad` constraints, whereas the non-CPS'd instances only have `Applicative` constraints).

**Comment by RyanGlScott** [comment]
Superseded by #119.
